### PR TITLE
Provide separate route tables for db/elasticache/redshift

### PR DESCRIPTION
**Pull request by murarisumit**
There is already [PR](https://github.com/terraform-aws-modules/terraform-aws-vpc/pull/122) for this. 

But I've added the suggestion you said up here.

Create three flag, set default to false.

* create_database_route_table
* create_redshift_route_table
* create_elasticache_route_table

If none flag is set, default behavior create private subnets.
Also added an example for it 

@antonbabenko please review.

### CODE REVIEW

File: examples/vpc-saperate-private-db-cache-route/main.tf [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  azs                 = ["ap-southeast-1a", "ap-southeast-1b"]
+  private_subnets     = ["10.10.1.0/24", "10.10.2.0/24"]
+  public_subnets      = ["10.10.11.0/24", "10.10.12.0/24"]
+  database_subnets    = ["10.10.21.0/24", "10.10.22.0/24"]
+  elasticache_subnets = ["10.10.31.0/24", "10.10.32.0/24"]
+  redshift_subnets    = ["10.10.41.0/24", "10.10.42.0/24"]
+
+  create_database_subnet_group   = false
+  create_database_route_table    = true
+  create_elasticache_route_table = true
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/main.tf]: Add `create_redshift_route_table = true`

File: examples/vpc-saperate-private-db-cache-route/main.tf [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+module "vpc" {
+  source = "./terraform-aws-vpc"
+
+  name = "sandbox"
+
+  cidr = "10.10.0.0/16"
+
+  azs                 = ["ap-southeast-1a", "ap-southeast-1b"]
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/main.tf]: Pick AZs from `eu-west-1`, please. All other examples are using that region. Have at least 3 AZs in examples.

File: examples/vpc-saperate-private-db-cache-route/main.tf [hunk 3]
```diff
@@ -0,0 +1,29 @@
+provider "aws" {
+  region = "eu-west-1"
+}
+
+module "vpc" {
+  source = "./terraform-aws-vpc"
+
+  name = "sandbox"
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/main.tf]: Call it `vpc-separate-private-route-tables`, also make sure you name directory the same.

File: examples/vpc-saperate-private-db-cache-route/main.tf [hunk 4]
```diff
@@ -0,0 +1,29 @@
+provider "aws" {
+  region = "eu-west-1"
+}
+
+module "vpc" {
+  source = "./terraform-aws-vpc"
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/main.tf]: `../../`

File: examples/vpc-saperate-private-db-cache-route/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# vpc saperate private db cache route table
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/README.md]: Please change to `# VPC with separate private route tables`.

Also link this example in `README.md` in the root.

File: examples/vpc-saperate-private-db-cache-route/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+Configuration in this directory creates set of VPC resources which may be sufficient for staging or production environment (look into [simple-vpc](../simple-vpc) for more simplified setup). 
+
+There are public, private, database, ElastiCache subnets, NAT Gateways created in each availability zone. **It sets up saperate private route for db, elasticache and redshift subnets.**.
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/README.md]: `separate`

File: examples/vpc-saperate-private-db-cache-route/main.tf [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+  create_database_subnet_group   = false
+  create_database_route_table    = true
+  create_elasticache_route_table = true
+  enable_nat_gateway             = true
+
+  tags = {
+    Owner       = "user"
+    Environment = "staging"
+    Name        = "complete"
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/main.tf]: `Name = "vpc-separate-private-route-tables"` - this is important to be able to find resources when running example.

File: examples/vpc-saperate-private-db-cache-route/main.tf [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  private_subnets     = ["10.10.1.0/24", "10.10.2.0/24"]
+  public_subnets      = ["10.10.11.0/24", "10.10.12.0/24"]
+  database_subnets    = ["10.10.21.0/24", "10.10.22.0/24"]
+  elasticache_subnets = ["10.10.31.0/24", "10.10.32.0/24"]
+  redshift_subnets    = ["10.10.41.0/24", "10.10.42.0/24"]
+
+  create_database_subnet_group   = false
+  create_database_route_table    = true
+  create_elasticache_route_table = true
+  enable_nat_gateway             = true
```
Comment (antonbabenko) [inline examples/vpc-saperate-private-db-cache-route/main.tf]: Also, `single_nat_gateway = true` is a good idea to save time when running example.

File: variables.tf [hunk 9]
```diff
@@ -46,6 +46,21 @@ variable "elasticache_subnets" {
   default     = []
 }
 
+variable "create_database_route_table" {
+  description = "Controls if saperate route table for database should be created"
```
Comment (antonbabenko) [inline variables.tf]: `separate` - check it everywhere. Also, I think `database subnet` is more correct - add word "subnet" to all variables.

File: examples/vpc-saperate-private-db-cache-route/main.tf (hunk 1 above)
Comment (murarisumit) [reply examples/vpc-saperate-private-db-cache-route/main.tf]: Done that

File: examples/vpc-saperate-private-db-cache-route/main.tf (hunk 2 above)
Comment (murarisumit) [reply examples/vpc-saperate-private-db-cache-route/main.tf]: Done, have used `eu-west-1` now and taken 3 AZs

File: examples/vpc-saperate-private-db-cache-route/main.tf (hunk 3 above)
Comment (murarisumit) [reply examples/vpc-saperate-private-db-cache-route/main.tf]: Done that

File: variables.tf (hunk 9 above)
Comment (murarisumit) [reply variables.tf]: Added this to all variables now.

**Comment by murarisumit** [comment]
@antonbabenko please review and provide feedback :)

**Comment by antonbabenko** [comment]
Hi @murarisumit ! Thank you for your contribution, but I won't be able to review it and release a new version during the next 30 days because I am on vacation in the middle of nowhere.

If you need this feature before that please use your fork in the meanwhile.

**Comment by antonbabenko** [comment]
Thanks @murarisumit!

[v1.38.0](https://github.com/terraform-aws-modules/terraform-aws-vpc/releases/v1.38.0) has been released.

**Pull request closed without merging**